File: mcp-code-intelligence-python/src/mcp_code_intel/orchestration/cache/invalidation.py
"""Cache invalidation — stale entry detection and eviction logic."""

from __future__ import annotations

import sys
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .cache_entry import CacheEntry

MAX_CACHE_SIZE = 10_000


def invalidate_stale(entries: list["CacheEntry"], current_hash: str) -> tuple[list["CacheEntry"], int]:
    """Remove entries with mismatched registry hash. Returns (kept, removed_count)."""
    kept = [e for e in entries if not e.is_stale(current_hash)]
    removed = len(entries) - len(kept)
    if removed > 0:
        _log(f"Invalidated {removed} stale entries (hash mismatch)")
    return kept, removed


def evict_lru(entries: list["CacheEntry"], max_size: int = MAX_CACHE_SIZE) -> list["CacheEntry"]:
    """Evict least-recently-used entries if over max_size."""
    if len(entries) <= max_size:
        return entries
    sorted_entries = sorted(entries, key=lambda e: e.last_hit, reverse=True)
    evicted = len(entries) - max_size
    _log(f"LRU eviction: removed {evicted} entries (size was {len(entries)})")
    return sorted_entries[:max_size]


def compute_token_overlap(query_tokens: set[str], entry_tokens: frozenset[str]) -> float:
    """Compute Jaccard-like overlap between query and entry tokens."""
    if not query_tokens or not entry_tokens:
        return 0.0
    intersection = query_tokens & entry_tokens
    union = query_tokens | entry_tokens
    return len(intersection) / len(union)


def _log(msg: str) -> None:
    print(f"[cache-invalidation] {msg}", file=sys.stderr, flush=True)

File: mcp-code-intelligence-nodejs/src/orchestration/cache/invalidation.ts
/**
 * Cache invalidation — stale entry detection and LRU eviction logic.
 */

import { CacheEntry, isStale } from './cache-entry.js';

const MAX_CACHE_SIZE = 10_000;

/** Remove entries with mismatched registry hash. Returns [kept, removedCount]. */
export function invalidateStale(entries: CacheEntry[], currentHash: string): [CacheEntry[], number] {
  const kept = entries.filter((e) => !isStale(e, currentHash));
  const removed = entries.length - kept.length;
  if (removed > 0) {
    console.error(`[cache-invalidation] Invalidated ${removed} stale entries`);
  }
  return [kept, removed];
}

/** Evict least-recently-used entries if over maxSize. */
export function evictLru(entries: CacheEntry[], maxSize: number = MAX_CACHE_SIZE): CacheEntry[] {
  if (entries.length <= maxSize) return entries;
  const sorted = [...entries].sort((a, b) => b.lastHit.localeCompare(a.lastHit));
  const evicted = entries.length - maxSize;
  console.error(`[cache-invalidation] LRU eviction: removed ${evicted} entries`);
  return sorted.slice(0, maxSize);
}

/** Compute Jaccard-like overlap between query tokens and entry tokens. */
export function computeTokenOverlap(queryTokens: Set<string>, entryTokens: Set<string>): number {
  if (queryTokens.size === 0 || entryTokens.size === 0) return 0;
  let intersection = 0;
  for (const t of queryTokens) {
    if (entryTokens.has(t)) intersection++;
  }
  const union = new Set([...queryTokens, ...entryTokens]).size;
  return intersection / union;
}
